File: Haskell/day2.hs
module Main where

import System.IO
import Text.Printf

main :: IO ()
main = do
  handle <- openFile "/mnt/c/Users/matth/Desktop/Advent Of Code 2021/Input/day2.txt" ReadMode
  contents <- hGetContents handle
  let inWords = words contents
  let directionPairs = map parseInput inWords
  printf "Part 1 solution: %s\n" (part1 directionPairs)
  printf "Part 2 solution: %s\n" (part2 directionPairs)
  hClose handle

parseInput :: String -> (Integer, Integer)
parseInput str
  | dir == "forward" = (scalar, 0)
  | dir == "down"    = (0, scalar)
  | dir == "up"      = (0, -scalar)
  | otherwise = (0, 0)
  where
    (dir, scalarStr) = break (== ' ') str
    scalar = read $ drop 1 scalarStr


part1 :: [(Integer, Integer)] -> String
part1 nums = show $ horiz * depth
  where
    (horiz, depth) = foldl (\(a, b) (u, v) -> (a + u, b + v)) (0, 0) nums

part2 :: [(Integer, Integer)] -> String
part2 nums = show $ horiz * depth
  where
    (nums1, nums2) = unzip nums
    tripleNums = zip3 nums1 nums2 (repeat 0)
    (aim, horiz, depth) = foldl (\(a, b, c) (u, v, w) -> (a + u, b + v, c + a * v)) (0, 0, 0) tripleNums

File: Rust/src/Day2.rs
const INPUT: &str = include_str!("..\\..\\Input\\day2.txt");

fn main() {
    let transformations: Vec<(i32, i32)> = INPUT
        .lines()
        .map(|x| x.split_at(x.find(' ').unwrap()))
        .map(|(d, s)| parse_input(d, s.trim()))
        .collect();
    part2(transformations);
}

fn part1(transformations: Vec<(i32, i32)>) {
    // Sum up all coordinate tuples
    let final_pos = transformations
        .iter()
        .fold((0, 0), |(u, v), (x, y)| (u + x, v + y));

    println!("Part 1 solution : {}", final_pos.0 * final_pos.1)
}

fn part2(transformations: Vec<(i32, i32)>) {
    let transformations3pair: Vec<(i32, i32, i32)> =
        transformations.iter().map(|(u, v)| (*u, *v, 0)).collect();

    // Tuple (aim, horizontal, depth)
    // Sum up all coordinate tuples
    let final_pos = transformations3pair
        .iter()
        .fold((0, 0, 0), |(u, v, w), (x, y, _)| {
            ((u + x), (v + y), (w + u * y))
        });

    println!("Part 2 solution : {}", final_pos.1 * final_pos.2)
}

// Tuple (depth, horizontal)
fn parse_input(direction: &str, scalar_str: &str) -> (i32, i32) {
    let scalar: i32 = scalar_str.parse().unwrap();
    match direction {
        "down" => (scalar, 0),
        "up" => (-scalar, 0),
        "forward" => (0, scalar),
        _other => (0, 0),
    }
}
